feat: add type hints to Client

rinarakaki opened the pull request:
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

rinarakaki committed:
add type hints

chalmerlowe commented on google/cloud/bigquery/client.py:
```diff
@@ -231,15 +230,23 @@ class Client(ClientWithProject):
+            Union[google.api_core.client_options.ClientOptions, Dict[str, Any]]
+        ] = None,
     ) -> None:
+        if client_options is None:
+            client_options = {}
+        if isinstance(client_options, dict):
+            client_options = google.api_core.client_options.from_dict(client_options)
+        assert isinstance(client_options, google.api_core.client_options.ClientOptions)
+
```
I am curious, why did we choose to:

* move this content from after the call to `super()` to before the call to `super()`? What is the reason for reordering this content?
* add an assert statement when one was not used previously?

rinarakaki replied:
- Moving this part before the call to `super()` is optional, but I did it because actually the exact same type check & coercion happens inside the `super()` (`Client.__init__`) as well, so the I let it early break in the second time.
- The reason for adding the assert statement is to let the type checker know the type of `client_options`, otherwise it wouldn't know that the `if client_options.api_endpoint:` expression at the line 261 is valid .

chalmerlowe commented:
@rinarakaki I ran the kokoro CI/CD checks.
There are a number of things that are going wrong both in the standard tests and in mypy.
I spent as much time as I could today trying to trouble shoot the issues, to no avail.
I have other tasks that require my attention. If you are able to figure out why these tests fail I am willing to come back and take another look at this.

#### mypy errors:
```
google/cloud/bigquery/client.py:265: error: Incompatible types in assignment
   (expression has type "str", target has type "Optional[ClientInfo]")  [assignment]
google/cloud/bigquery/client.py:269: error: Incompatible types in assignment
   (expression has type "str", target has type "Optional[ClientInfo]")  [assignment]
```

#### unit test errors:
```ImportError while loading conftest '/tmpfs/src/github/python-bigquery/tests/unit/conftest.py'.
tests/unit/conftest.py:19: in 
    from .helpers import make_client
tests/unit/helpers.py:19: in 
    import google.cloud.bigquery.client
google/cloud/bigquery/__init__.py:36: in 
    from google.cloud.bigquery.client import Client
google/cloud/bigquery/client.py:185: in 
    class Client(ClientWithProject):
google/cloud/bigquery/client.py:235: in Client
    _http: Optional[requests.Session] = None,
E   NameError: name 'requests' is not defined
```

chalmerlowe committed:
Merge branch 'main' into typing-client

rinarakaki committed:
Merge branch 'main' into typing-client

chalmerlowe committed:
Update client.py

Moves import from being used solely during specific checks to being more universally available.

chalmerlowe committed:
Merge branch 'main' into typing-client

chalmerlowe commented on google/cloud/bigquery/client.py:
```diff
@@ -133,8 +134,7 @@
     PathType = Union[str, bytes, os.PathLike[str], os.PathLike[bytes]]
-    import requests  # required by api-core
-
+    
```
```suggestion
```

chalmerlowe committed:
Update google/cloud/bigquery/client.py

chalmerlowe committed:
Update client.py

testing some minor changes to deal with mypy quirks

chalmerlowe commented on google/cloud/bigquery/client.py:
```diff
@@ -122,9 +125,7 @@
-pandas = (
-    _versions_helpers.PANDAS_VERSIONS.try_import()
-)  # mypy check fails because pandas import is outside module, there are type: ignore comments related to this
+pandas = _versions_helpers.PANDAS_VERSIONS.try_import()  # mypy check fails because pandas import is outside module, there are type: ignore comments related to this
```
```suggestion
pandas = (
    _versions_helpers.PANDAS_VERSIONS.try_import()
)  # mypy check fails because pandas import is outside module, there are type: ignore comments related to this

```

chalmerlowe committed:
Update google/cloud/bigquery/client.py

chalmerlowe reviewed (approved):
LGTM
